src/proto/h1/dispatch.rs

```diff
@@ -180,12 +186,31 @@ where
         task::yield_now(cx).map(|never| match never {})
     }
 
+    fn check_header_read_timeout(&mut self, cx: &mut task::Context<'_>) -> crate::Result<()> {
+        if let Some(header_read_timeout) = self.header_read_timeout {
+            if let Some(read_timeout_fut) = &mut self.header_read_timeout_fut {
+                if Pin::new(read_timeout_fut).poll(cx).is_ready() {
+                    warn!("read header from client timeout");
+                    return Err(crate::Error::new_header_timeout());
+                }
+            } else {
+                self.header_read_timeout_fut = Some(Box::pin(sleep(header_read_timeout)))
+            }
+        }
+
+        Ok(())
+    }
+
     fn poll_read(&mut self, cx: &mut task::Context<'_>) -> Poll<crate::Result<()>> {
         loop {
             if self.is_closing {
                 return Poll::Ready(Ok(()));
             } else if self.conn.can_read_head() {
-                ready!(self.poll_read_head(cx))?;
+                if let Poll::Ready(read_head_rt) = self.poll_read_head(cx) {
+                    read_head_rt?;
+                }
+                self.check_header_read_timeout(cx)?;
+                self.header_read_timeout_fut = None;
```

Comment:
I believe your last change invalidated the timeout :smiley: 

```suggestion
                if let Poll::Ready(read_head_rt) = self.poll_read_head(cx) {
                    self.header_read_timeout_fut = None;
                    read_head_rt?;
                }
                self.check_header_read_timeout(cx)?;
```